syntax = "proto3";
package twitter;

import "validator.proto";


message Item {
  int64 id = 1;
  string permalink = 2;
  int64 timestamp = 3;
  string created_at = 4;
  string created_at_time = 5;
  string full_text_norm = 6;
  string full_text_clean = 7;
  string full_text = 8;
  string text_sentiment = 9;
  string user_id = 10;
  string user_name = 11;
  string screen_name = 12;
  string user_description = 13;
  bool user_verified = 14;
  string user_created_at = 15;
  int32 user_followers_count = 16;
  int32 user_friends_count = 17;
  int32 user_statuses_count = 18;
  int32 user_favourites_count = 19;
  int32 user_listed_count = 20;
  int32 user_media_count = 21;
  int32 user_location = 22;
  bool user_protected = 23;
  string user_link_url = 24;
  string user_profile_image_url = 25;
  int32 reply_count = 26;
  int32 retweet_count = 27;
  int32 like_count = 28;
  int32 quote_count = 29;
  int32 conversation_id = 30;
  string device = 31;
  string lang = 32;
  repeated mediaItem media = 33;
  repeated Mention mentions = 34;
  repeated Hashtag hashtags = 35;
  int64 engagement = 36;
  int64 potential_reach = 37;
  float engagement_rate = 38;
}

message mediaItem {
  string type = 1;
  repeated Source source = 2;
}

message Source {
  string preview_url = 1;
  string full_url = 2;
}

message Mention {
  string text = 1;
  string display_name = 2;
}

message Hashtag {
  string text = 1;
}


service twitter {
  rpc StreamRequest (twitterRequest) returns (twitterResponse) {};
}

message twitterRequest {
  string keyword = 1 [(validator.field) = {string_not_empty: True}];
  string search_type = 2 [(validator.field) = {string_not_empty: True}];
  string since = 3 [(validator.field) = {string_not_empty: True}];
  string until = 4 [(validator.field) = {string_not_empty: True}];
}

message twitterResponse {
  string message = 1;
  string updateAt = 2;
  repeated Item items = 3;
}

